rules/otbn.bzl

```diff
@@ -148,6 +148,45 @@ def _otbn_binary(ctx):
         ),
     ]
 
+def _otbn_test(ctx):
+    """This rule is for standalone OTBN unit tests, which are run on the host
+    via the OTBN simulator.
+
+    It first generates binaries using the same method as otbn_binary, then runs
+    them on the simulator. Tests are expected to count failures in the w0
+    register; the test checks that w0=0 to determine if the test passed.
+    """
+    providers = _otbn_binary(ctx)
+
+    # Extract the output .elf file from the output group.
+    elf = providers[1].elf.to_list()[0]
+
+    # Create a simple script that runs the OTBN simulator on the .elf file and
+    # checks if the w0 register is 0.
+    simulator_cmd = "{} {} --dump-regs -".format(ctx.file._simulator.path, elf.short_path)
+    expected_string = "0x0000000000000000000000000000000000000000000000000000000000000000"
+    script = "\n".join([
+        'echo "Running simulator: {}"'.format(simulator_cmd),
+        'result=$({} | grep -m 1 -P "w0  = [0-9A-Fa-f]+")'.format(simulator_cmd),
+        'echo "Got     : $result"',
+        'echo "Expected:  w0  = {}"'.format(expected_string),
+        'if [[ "$result" == *"{}" ]]; then'.format(expected_string),
+        '  echo "PASS"',
+        "  exit 0",
+        "fi",
+        'echo "FAIL"',
+        "exit 1",
+    ])
+    ctx.actions.write(
+        output = ctx.outputs.executable,
```

Comment:
Bazel actually recommends declaring a new file with `ctx.actions.declare_file` and making it the `DefaultInfo`'s `executable` field. `ctx.outputs.executable` is something of a legacy thing.

Reply:
Thanks, fixed!